metric_test.go

```diff
@@ -47,3 +49,81 @@ func TestTagValidation(t *testing.T) {
 		}
 	}
 }
+
+func TestNameWithTags(t *testing.T) {
+	type testCase struct {
+		expectedName         string
+		expectedNameWithTags string
+		expectedTags         []string
+		md                   MetricDefinition
+	}
+
+	testCases := []testCase{
+		{
+			"a.b.c",
+			"a.b.c;tag1=value1",
+			[]string{"tag1=value1"},
+			*NewMetricDefinition("a.b.c", []string{"tag1=value1", "name=ccc"}),
+		}, {
+			"a.b.c",
+			"a.b.c;a=a;b=b;c=c",
+			[]string{"a=a", "b=b", "c=c"},
+			*NewMetricDefinition("a.b.c", []string{"name=a.b.c", "c=c", "b=b", "a=a"}),
+		}, {
+			"a.b.c",
+			"a.b.c",
+			[]string{},
+			*NewMetricDefinition("a.b.c", []string{"name=a.b.c"}),
+		}, {
+			"a.b.c",
+			"a.b.c",
+			[]string{},
+			*NewMetricDefinition("a.b.c", []string{}),
+		}, {
+			"c",
+			"c;a=a;b=b;c=c",
+			[]string{"a=a", "b=b", "c=c"},
+			*NewMetricDefinition("c", []string{"c=c", "a=a", "b=b"}),
+		},
+	}
+
+	for _, tc := range testCases {
+		tc.md.SetId()
+		if tc.expectedName != tc.md.Name {
+			t.Fatalf("Expected name %s, but got %s", tc.expectedName, tc.md.Name)
+		}
+
+		if tc.expectedNameWithTags != tc.md.NameWithTags {
+			t.Fatalf("Expected name with tags %s, but got %s", tc.expectedNameWithTags, tc.md.NameWithTags)
+		}
+
+		if len(tc.expectedTags) != len(tc.md.Tags) {
+			t.Fatalf("Expected tags %+v, but got %+v", tc.expectedTags, tc.md.Tags)
+		}
+
+		for i := range tc.expectedTags {
+			if len(tc.expectedTags[i]) != len(tc.md.Tags[i]) {
+				t.Fatalf("Expected tags %+v, but got %+v", tc.expectedTags, tc.md.Tags)
+			}
+		}
+
+		getAddress := func(s string) uint {
+			return uint((*reflect.StringHeader)(unsafe.Pointer(&s)).Data)
```

Comment:
what's the purpose of uintptr -> uint ?

Reply:
we'll want to use the address like a numeric value, f.e. when we add the length of `NameWithTags` a few lines later.